pkg/hubble/exporter/aggregator.go

```diff
@@ -0,0 +1,187 @@
+	}
+}
+
+// NewAggregatorWithFields creates a new aggregator with field aggregation
+func NewAggregatorWithFields(fieldAggregate fa.FieldAggregate, logger *slog.Logger) *Aggregator {
+	return &Aggregator{
+		m:              make(map[AggregateKey]*AggregateValue),
+		fieldAggregate: fieldAggregate,
+		logger:         logger.With(logfields.LogSubsys, "hubble-aggregator"),
+	}
+}
+
+func (a *Aggregator) Add(ev *v1.Event) {
+	f := ev.GetFlow()
+	if f == nil {
+		return
+	}
+
+	processedFlow := &flowpb.Flow{}
+	a.fieldAggregate.Copy(processedFlow.ProtoReflect(), f.ProtoReflect())
+
+	k := a.generateAggregationKey(processedFlow)
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	v, ok := a.m[k]
+	if !ok {
+		switch f.GetTrafficDirection() {
+		case flowpb.TrafficDirection_INGRESS:
+			v = &AggregateValue{
+				IngressFlowCount: 1,
+				ProcessedFlow:    processedFlow,
+			}
+		case flowpb.TrafficDirection_EGRESS:
+			v = &AggregateValue{
+				EgressFlowCount: 1,
+				ProcessedFlow:   processedFlow,
+			}
+		default:
+			v = &AggregateValue{
+				UnknownDirectionFlowCount: 1,
+				ProcessedFlow:             processedFlow,
+			}
+		}
+		a.m[k] = v
+	} else {
+		switch f.GetTrafficDirection() {
+		case flowpb.TrafficDirection_INGRESS:
+			v.IngressFlowCount++
+		case flowpb.TrafficDirection_EGRESS:
+			v.EgressFlowCount++
+		default:
+			v.UnknownDirectionFlowCount++
+		}
+	}
+}
+
+func (a *Aggregator) generateAggregationKey(processedFlow *flowpb.Flow) AggregateKey {
+	b, _ := proto.Marshal(processedFlow.ProtoReflect().Interface())
+	return AggregateKey(b)
+}
+
+// AggregatorRunner manages the lifecycle of an aggregator with periodic export.
+type AggregatorRunner struct {
+	aggregator *Aggregator
+	interval   time.Duration
+	encoder    Encoder
+	logger     *slog.Logger
+
+	stop chan struct{}
+	wg   sync.WaitGroup
+}
+
+func (r *AggregatorRunner) Start() {
+	r.stop = make(chan struct{})
+	r.wg.Add(1)
+	go r.run()
+}
+
+func (r *AggregatorRunner) Stop() {
+	if r.stop != nil {
+		close(r.stop)
+	}
+	r.wg.Wait()
+	r.stop = nil
+}
+
+func (r *AggregatorRunner) Add(event *v1.Event) {
+	r.aggregator.Add(event)
+}
+
+func (r *AggregatorRunner) run() {
+	defer r.wg.Done()
+	ticker := time.NewTicker(r.interval)
+
+	for {
+		select {
+		case <-r.stop:
+			// Flush before stopping.
+			r.exportAggregates()
+			return
+		case <-ticker.C:
+			r.exportAggregates()
+		}
+	}
+}
+
+func (r *AggregatorRunner) exportAggregates() {
+	r.aggregator.mu.Lock()
+	defer r.aggregator.mu.Unlock()
+
+	for _, value := range r.aggregator.m {
+		exportEvent := processedFlowToAggregatedExportEvent(value.ProcessedFlow, value.IngressFlowCount, value.EgressFlowCount, value.UnknownDirectionFlowCount)
+		if exportEvent == nil {
+			continue
+		}
+
+		if err := r.encoder.Encode(exportEvent); err != nil {
+			r.logger.Error("Failed to export aggregate", logfields.Error, err)
+		}
+	}
+
+	r.aggregator.m = make(map[AggregateKey]*AggregateValue)
+}
```

Comment:
suggestion: Seems like this method only acts on aggregator fields, it could probably be moved to the `Aggregator` and passed the encoder. That way `AggregatorRunner` doesn't need to know about the aggregator internals.